Add/newsletter offline

millerf opened the pull request:
New take on https://github.com/Automattic/jetpack/pull/27144

Works alongside D88449-code#change-YBbSEaQmNldl

millerf committed:
[not verified] add user_can_receive_post_by_mail

millerf committed:
Updated visitor_can_view_content to user visitor_id && Added user_can_receive_post_by_mail

millerf commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-offline-subscription-service.php:
```diff
@@ -52,6 +57,41 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+	 */
+	public static function user_can_receive_post_by_mail( $user_id, $blog_id, $post_id ) {
+		if ( ! ( defined( 'IS_WPCOM' ) && IS_WPCOM ) ) {
+			throw \Exception( 'Should only be called on WPCOM' );
```
I am not sure about that, but I wanted to add it either way, as the code won't work outside of WPCOM

artpi replied:
I don't think we need that

millerf commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-offline-subscription-service.php:
```diff
@@ -52,6 +57,41 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+
+		if ( empty( $access_level ) || $access_level === 'everybody' ) {
+			// empty level means the post is not gated for paid users
+			return true;
```
I was not sure if that logic would be included in `visitor_can_view_content`. It's a good shortcut for now.

millerf commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-subscription-service.php:
```diff
@@ -43,10 +43,11 @@ public function initialize();
 	 * @return boolean
 	 */
-	public function visitor_can_view_content( $valid_plan_ids, $access_level );
+	public function visitor_can_view_content( $valid_plan_ids, $access_level, $visitor_id = null );
```
I had to change the signature again as it is used outside of request context (async-jobs), and therefore does not have a current_user

artpi replied:
You should `wp_set_current_user` and it will be fine. Plenty of jobs do that

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-offline-subscription-service.php:
```diff
@@ -52,6 +57,41 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+			throw \Exception( 'Should only be called on WPCOM' );
+		}
+
+		// Site admins can do everything
```
Why don't we just do something like
```
$previous_user = wp_get_current_user();
wp_set_current_user( $user_id );
return $this->visitor_can_view_content(  [],  get_post_meta( $post_id, '_jetpack_newsletter_access', true ) );
wp_set_current_user( $previous_user );

```

millerf replied:
Lol

millerf committed:
Artur's feedback

- Remove $visitor_id parameter
- Remove IS_WPCOM check

millerf committed:
Rework the way access is granted and restructure to centralize the logic

millerf commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,43 +71,43 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+			$is_paid_subscriber = $this->validate_subscriptions( $valid_plan_ids, $subscriptions );
+		} else {
+			// Token not valid. We bail even unless the post can be accessed publicly.
+			return 'everybody' === $access_level;
```
Still not exactly sure about that. What is the use case for the token not to be valid...

artpi replied:
1. If user is logged in AND can edit post, than also unlock (I forgot about the use case)
2. If access_level is empty, than also unlock

artpi replied:
I think you can just don't have this check here and forward to user_has_access( $access_level, false, false )

artpi replied:
But you will need to put `current_user_can('edit_post')`

millerf replied:
Done

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,43 +71,43 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+	 *
+	 * @return bool does the user have access to the post/blog.
+	 */
+	protected static function user_has_access( $access_level, $is_blog_subscrber, $is_paid_subscriber ) {
```
Yes fantastic, this method needs to be broken out 👍

artpi replied:
Are we in the most base class? We should have this in the most base class in this inheritance chain

millerf replied:
Yes, all the parents are interfaces

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-offline-subscription-service.php:
```diff
@@ -52,6 +53,57 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+			$blog_id
+		);
+
+		$access_level       = get_post_meta( $post_id, '_jetpack_newsletter_access', true );
```
Would be sweet to break it out as well into a method

millerf replied:
Now that I reused `visitor_can_view_content()` ( see https://github.com/Automattic/jetpack/pull/27352/files#r1021249651), I am not sure it is still worth it...

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-offline-subscription-service.php:
```diff
@@ -52,6 +53,57 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+		 * Filter the subscriptions attached to a specific user on a given site.
+		 *
+		 * @since 9.4.0
+		 *
```
I think we should put some of this content in the `visitor_can_view_content` in this class.
I imagine WPCOM_Offline_Subscription_Service being used in email and reader and visitor_can_view_content is probably reader

millerf replied:
Ok

millerf replied:
I used `WPCOM_Offline_Subscription_Service::visitor_can_view_content()` here. I had overlooked that method.

millerf committed:
Feedback https://github.com/Automattic/jetpack/pull/27352/files#r1020302640

millerf committed:
Feedback https://github.com/Automattic/jetpack/pull/27352/files#r1020312430

millerf committed:
Make get_site_id static

millerf committed:
Removed over engineered get_site_id()

millerf committed:
add filter test_jetpack_is_supported_jetpack_recurring_payments for is_supported_jetpack_recurring_payments()

millerf committed:
Make methods not static

millerf committed:
Added tests

millerf committed:
Add post_id to prevent deprecation notice

millerf committed:
Fix tests

millerf committed:
Some clean up

millerf committed:
Some refactor

millerf committed:
[not verified] fix compatibility issue

millerf committed:
[not verified] fix tests

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,43 +71,50 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
-			 * get current subscriptions and determine if the user has
-			 * a valid subscription to match the plan ID.
-			 */
+			$is_blog_subscrber  = 'active' === $payload['blog_sub'];
```
`subscriber`

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,43 +71,50 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+	 *
+	 * @return bool does the user have access to the post/blog.
+	 */
+	protected function user_has_access( $access_level, $is_blog_subscrber, $is_paid_subscriber, $post_id ) {
```
I wonder if this should be public and static, allowing for easier use.
ideally `Subscription_Service` should be a class and not an interface and this method would be there, but that may be too disruptive in this PR

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -153,7 +153,7 @@ public function access_url( $mode = 'subscribe' ) {
 		}
 
-		$login_url = $this->get_rest_api_token_url( $this->get_site_id(), $permalink );
+		$login_url = $this->get_rest_api_token_url( get_current_blog_id(), $permalink );
```
This change is problematic because on Jetpack `get_current_blog_id` has a different meaning and we want to get the WPCOM blog id.
Jetpack site may be a multisite as well and you will just get the blog id of the Jetpack site here

millerf replied:
I don't understand that comment...

> on Jetpack get_current_blog_id has a different meaning and we want to get the WPCOM blog id.

What's the "Jetpack meaning"?

At the end of the day, `$this->get_site_id()` was calling `get_current_blog_id()` so there is no difference here, is there?

artpi replied:
Here is a case you missed:
https://github.com/Automattic/jetpack/blob/b313bd253afdf919adf7a71a866d0ab85b3a7bda/projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-jetpack-token-subscription-service.php#L35

artpi replied:
Jetpack stores the id of the connected shadow site as an option. Because that ID refers to WPCOM id

millerf replied:
I see!!!

millerf committed:
Fix typo

millerf committed:
[not verified] Refactor

millerf committed:
Remove unecessary $blog_id

millerf committed:
[not verified] Fix tests

millerf committed:
Fix typo

millerf committed:
Reinstate get_site_id()

millerf committed:
[not verified] Added some test for online

millerf committed:
Fix JWT encode routine

millerf committed:
More fix for the tests

millerf committed:
[not verified] Test with admin user

millerf committed:
Small changes

millerf committed:
Added a skipped test for cookies

romarioraffington commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-wpcom-offline-subscription-service.php:
```diff
@@ -44,12 +52,35 @@ public static function available() {
+		$subscriptions      = self::abbreviate_subscriptions( $subscriptions );
+		$is_paid_subscriber = $this->validate_subscriptions( $valid_plan_ids, $subscriptions );
+
+		return $this->user_has_access( $access_level, true, $is_paid_subscriber, $this->post_id );
```
It would be nice make `true` a variable for ease of readability

millerf replied:
You should use PHPStorm ;)

<img width="949" alt="Screenshot 2022-11-18 at 11 41 27" src="https://user-images.githubusercontent.com/790558/202693763-909f22d4-f9e4-41fd-9328-afd5ae2ba7dd.png">

artpi replied:
There is a plugin for VS Code:
https://marketplace.visualstudio.com/items?itemName=BobbyZrncev.intellij-parameter-hints

romarioraffington commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,43 +71,50 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
+		if ( $access_level === 'paid_subscribers' ) {
+			return $is_paid_subscriber;
 		}
 
```
Should we change this to Yoda conditions?

millerf replied:
For some reason my IDE yells when I put Yoda conditions...

artpi replied:
Your IDE is misconfigured then. Check if it is not configured for WooCommerce. You should be using Jetpack coding standards

millerf replied:
I'll have a look

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-jwt.php:
```diff
@@ -193,12 +193,12 @@ public static function encode( $payload, $key, $alg = 'HS256', $key_id = null, $
 		$segments      = array();
-		$segments[]    = static::urlsafe_b64_decode( static::json_encode( $header ) );
-		$segments[]    = static::urlsafe_b64_decode( static::json_encode( $payload ) );
+		$segments[]    = static::urlsafe_b64_encode( static::json_encode( $header ) );
```
This class is widely used and it makes me suspicious if there is such a blatant error

millerf replied:
It is only used to decode and not to encode...
I foud the issue by replacing with the WPCOM class equivalent and inspecting the differences (it is mainly a lot of lint changes), adding a new cypher algorythm AND using encode here instead of decode...

Have a look at that link: fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Syvo%2Swjg%2Spynff.wjg.cuc%3Se%3Qpn0osq59%23159%2Q165-og , it is makes things obvious.

artpi replied:
This is all a nice find.
Can you please create a release note for that?

millerf replied:
Done

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -177,6 +177,10 @@ private function token_from_cookie() {
 	 * @return void
 	 */
 	private function set_token_cookie( $token ) {
+		if ( defined( 'TESTING_IN_JETPACK' ) && TESTING_IN_JETPACK ) {
```
Why so?

millerf replied:
Otherwise I get a 
```
Warning: Cannot modify header information - headers already sent (output started at script:line[...])
```
Headers are sent in bootstrap.php

artpi replied:
I see

artpi commented on projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-token-subscription-service.php:
```diff
@@ -71,43 +71,50 @@ public function visitor_can_view_content( $valid_plan_ids, $access_level ) {
-			 * @param int   $user_id The user's ID.
-			 * @param int   $site_id ID of the current site.
-			 */
-			$subscriptions = apply_filters(
```
I think this was to accomodate the scenario of:

Scenario | Site type | LoggedIn to WPCOM? | Has proper JWT Cookie? | Subscribed as blog subscriber? | Paid subscriber? | Post-level visibility setting | Subscriber email sent? | User sees the post | What happens in the modal | Notes
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
A | WPCOM Simple, *.wordress.com domain | ✅ | ✅ / ❌ | ✅ | ✅ | Paid-Subscriber | ✅ | ✅ | – | User has access due to to WPCOM logged-in cookie and being a subscriber

Are we still handling this case?

millerf replied:
Should that case be handled by Token_Subscription_Service? Maybe it should be somewhere else... Also I am pretty sure you made me remove that part 🤔

artpi replied:
I think it should be moved to `Jetpack_Token_Subscription_Service` - does this make sense?

artpi replied:
It depends which one ultimately controls displaying on WPCOM logged-in page frankly. We may need to look at the code that switches to appropriate service. With the move from FSE to jetpack and now jetpack being on wpcom this whole logic may have become messed up

millerf replied:
Do you know where this is/will be called? I am still struggling to see the PCOM use-case. For me WPCOM was the email/Offline service. And for the rest we will eventually only use the token...

artpi commented on projects/plugins/jetpack/modules/memberships/class-jetpack-memberships.php:
```diff
@@ -482,7 +479,10 @@ public static function is_enabled_jetpack_recurring_payments() {
 	public static function is_supported_jetpack_recurring_payments() {
-		return (
+
+		$is_supported_jetpack_recurring_payments = apply_filters( 'test_jetpack_is_supported_jetpack_recurring_payments', false );
```
```

$api_available = ( ( defined( 'IS_WPCOM' ) && IS_WPCOM ) || Jetpack::is_connection_ready() );
$supported_in plan = Jetpack_Plan::supports( 'recurring-payments' );
return apply_filters( 'test_jetpack_is_supported_jetpack_recurring_payments', (  $api_available &&  $supported_in plan ) );
 
```

artpi commented on projects/plugins/jetpack/tests/php/modules/subscriptions/test_class.jetpack-subscriptions-offline.php:
```diff
@@ -0,0 +1,73 @@
+	public function test_jetpack_paid_newsletters_non_gated_post() {
+		$post_id = $this->setup_jetpack_paid_newsletters();
+
+		$subscription_service = new WPCOM_Offline_Subscription_Service();
```
We are testing WPCOM_Offline_Subscription_Service a lot here, and we also should be testing all the other services

millerf replied:
I am testing the Token service in the other file... 
Should we test all the other ones as well?

artpi replied:
Yeah, I see.
But you will need to write similar tests for `Jetpack_Token_Subscription_Service` and cover it there

millerf replied:
But they are tested... 🤔

artpi commented on projects/plugins/jetpack/tests/php/modules/subscriptions/test_class.jetpack-subscriptions-offline.php:
```diff
@@ -0,0 +1,73 @@
+		$subscription_service = new WPCOM_Offline_Subscription_Service();
+		$this->assertTrue( $subscription_service->subscriber_can_receive_post_by_mail( $this->regular_subscriber_id, $post_id ) );
+		$this->assertTrue( $subscription_service->subscriber_can_receive_post_by_mail( $this->paid_subscriber_id, $post_id ) );
+		$this->assertTrue( $subscription_service->subscriber_can_receive_post_by_mail( $this->admin_user_id, $post_id ) );
```
Let's add assertTrue for user id `0` (guest)

in fact, all of the below cases need that

millerf replied:
This is only used in the async-job to send emails. We have to assume that:
- user passed IS subscriber
- user passed IS (obviously) not guest

artpi replied:
👍

artpi commented on projects/plugins/jetpack/tests/php/modules/subscriptions/test_class.jetpack-subscriptions-offline.php:
```diff
@@ -0,0 +1,73 @@
+		// All subscribers should see the post
+		$this->assertTrue( $subscription_service->subscriber_can_receive_post_by_mail( $this->regular_subscriber_id, $post_id ) );
+		$this->assertTrue( $subscription_service->subscriber_can_receive_post_by_mail( $this->paid_subscriber_id, $post_id ) );
+		$this->assertTrue( $subscription_service->subscriber_can_receive_post_by_mail( $this->admin_user_id, $post_id ) );
```
assertFalse for user id 0

millerf replied:
Same than before. This use case cannot happen if offline.

artpi replied:
👍

artpi commented on projects/plugins/jetpack/tests/php/modules/subscriptions/test_class.jetpack-subscriptions-online.php:
```diff
@@ -0,0 +1,251 @@
+	/**
+	 * Jetpack regular user.
+	 */
+	public function test_jetpack_regular_user() {
```
Let's call the user a non-subscriber maybe?

millerf replied:
Sure... why not...

millerf replied:
Done

artpi commented on projects/plugins/jetpack/tests/php/modules/subscriptions/test_class.jetpack-subscriptions-online.php:
```diff
@@ -0,0 +1,251 @@
+		$GLOBALS['post'] = get_post( $post_id );
+
+		$previous_user = wp_get_current_user();
+		wp_set_current_user( 0 );
```
Nice!

artpi reviewed (commented):
NICE! I left some comments, code looks good.
Now I need to test it all

millerf committed:
Added changelog entry

millerf committed:
Rewrite is_supported_jetpack_recurring_payments()

millerf committed:
Change test name

millerf committed:
Rename and refactor most of the logic for WPCOM online

- Rename  WPCOM_Offline_Subscription_Service to WPCOM_Online_Subscription_Service
- Available across all WPCOM as soon as a user is loggued in
- Added front-end use case
- Added tests for front-end
